File: cpp/medium/LongestCommonSubsequence.cpp
#include <algorithm>
#include <string>
#include <vector>
using namespace std;

int longestCommonSubsequence(string text1, string text2) {
  vector<vector<int>> dp(text1.size() + 1, vector(text2.size() + 1, 0));
  int rs = 0;
  for (int i = 1; i <= text1.size(); ++i) {
    for (int k = 1; k <= text2.size(); ++k) {
      if (text1[i - 1] == text2[k - 1]) {
        dp[i][k] = dp[i - 1][k - 1] + 1;
      } else {
        dp[i][k] = max(dp[i - 1][k], dp[i][k - 1]);
      }
      if (dp[i][k] > rs) {
        rs = dp[i][k];
      }
    }
  }
  return rs;
}

File: go/middle/longestCommonSubsequence.go
package main

func longestCommonSubsequence(text1 string, text2 string) int {
	dp := make([][]int, len(text1)+1)
	for i := range dp {
		dp[i] = make([]int, len(text2)+1)
	}
	for i := 1; i < len(dp); i++ {
		for k := 1; k < len(dp[i]); k++ {
			if text1[i-1] == text2[k-1] {
				dp[i][k] = dp[i-1][k-1] + 1
			} else {
				dp[i][k] = max(dp[i-1][k], dp[i][k-1])
			}
		}
	}
	return dp[len(dp)-1][len(dp[0])-1]
}
